lib/document.js

```diff
@@ -580,17 +580,12 @@ Document.prototype._batchSave = function(executeInsert) {
  * @return {Promise}
  */
 Document.prototype._batchSaveSelf = function(executeInsert) {
-  var model = this._getModel();
+  var self = this;
 
   return new Promise(function(resolve, reject) {
-    if (model._tableReady === true) {
+    self.constructor.ready().then(function() {
```

Comment:
One more comment: To be clear this work, but the prototype chain is tricky, and it's better for future generation to call `ready` exactly on the appropriate object so hopefully people will be less confuse if they try to understand things work :)